README.md
# twutil

A set of utilities for handling Twitter data.

## Installing

`pip install twutil`

Twitter API calls rely on the [TwitterAPI](https://github.com/geduldig/TwitterAPI) Python library.

To use OAuth, you'll need to set the following environmental variables:

`TW_CONSUMER_KEY`
`TW_CONSUMER_SECRET`
`TW_ACCESS_TOKEN`
`TW_ACCESS_TOKEN_SECRET`







HISTORY.rst
.. :changelog:

History
-------
0.2.2 (2015-10-08)
++++++++++++++++++
* add get_friends

0.2.1 (2015-10-06)
++++++++++++++++++
* Properly specify requirements.

0.2.0 (2015-10-06)
++++++++++++++++++
* Backward compatibility for python2


0.1.9 (2014-12-08)
++++++++++++++++++
* Convert to Python3
* Support to collect friends

0.1.8 (2014-12-08)
++++++++++++++++++
* Get members of a Twitter list
* Improvements to tokenizer.
* Lookup screen names in bulk.


0.1.7 (2014-05-15)
++++++++++++++++++
* Fetch followers
* Enhanced error handling
* Bug fixes

0.1.6 (2014-05-02)
++++++++++++++++++
* Update versions


0.1.5 (2014-05-02)
++++++++++++++++++
* Bug fixes
* Skip missing users

0.1.4 (2014-05-02)
++++++++++++++++++
* Bug fixes
* Fail more gracefully

0.1.3 (2014-05-01)
++++++++++++++++++
* Small doc changes


0.1.2 (2014-05-01)
++++++++++++++++++

* Switch to TwitterAPI
* Support search and track by user
* rename api module to collect

0.1.0 (2014-03-29)
++++++++++++++++++

* First release on PyPI.


tests/test_twutil.py
#!/usr/bin/env python
# -*- coding: utf-8 -*-

"""
test_twutil
----------------------------------

Tests for `twutil` module.
"""

import unittest

from twutil import twutil


class TestTwutil(unittest.TestCase):

    def setUp(self):
        pass

    def test_something(self):
        pass

    def tearDown(self):
        pass

if __name__ == '__main__':
    unittest.main()

twutil/api.py
import os
from Queue import Queue
import sys
from threading import Thread
import time
import twitter

twapi = twitter.Twitter(auth=twitter.OAuth(os.environ.get('TW_ACCESS_TOKEN'),
                                           os.environ.get('TW_ACCESS_TOKEN_SECRET'),
                                           os.environ.get('TW_CONSUMER_KEY'),
                                           os.environ.get('TW_CONSUMER_SECRET')))

twstream = twitter.TwitterStream(auth=twitter.OAuth(os.environ.get('TW_ACCESS_TOKEN'),
                                                    os.environ.get('TW_ACCESS_TOKEN_SECRET'),
                                                    os.environ.get('TW_CONSUMER_KEY'),
                                                    os.environ.get('TW_CONSUMER_SECRET')))


def tweets_for_user(screen_name, limit=1e10):
    qu = Queue()
    p = Thread(target=_tweets_for_user, args=(qu, screen_name, limit))
    p.start()
    p.join(900)
    if p.is_alive():
        sys.stderr.write('no results after 15 minutes for %s. Aborting.' % screen_name)
        return None
    else:
        return qu.get()


def _tweets_for_user(qu, screen_name, limit=1e10):
    max_id = None
    tweets = []
    while True:
        try:
            if max_id:
                response = twapi.statuses.user_timeline(screen_name=screen_name, count=200, max_id=max_id)
            else:
                response = twapi.statuses.user_timeline(screen_name=screen_name, count=200)
            if len(response) == 0:
                qu.put(tweets)
                return
            else:
                sys.stderr.write('fetched %d more tweets for %s\n' % (len(response), screen_name))
                tweets.extend(response)
                if len(tweets) >= limit:
                    qu.put(tweets[:limit])
                    return
                # print json.dumps(response)
                max_id = min(t['id'] for t in response) - 1
        except Exception as e:
            if '404' in str(e.e):
                sys.stderr.write('Error: %s\nSkipping...\n' % e)
                qu.put(None)
                return
            else:
                sys.stderr.write('Error: %s\nSleeping for 5 minutes...\n' % e)
                time.sleep(300)


twutil/data.py
import json


class Tweet(object):
    """
    >>> tweet = Tweet(json.loads(test_tweet))
    >>> tweet.js['user']['name']
    u'Ellen DeGeneres'
    """

    def __init__(self, js):
        self.js = js

    def __str__(self):
        return u'\t'.join([self.js['user']['screen_name'], self.js['created_at'], self.js['text']])


def jsons2tweets(jsons):
    """
    >>> tweets = [t for t in jsons2tweets([test_tweet, test_tweet])]
    >>> len(tweets)
    2
    >>> tweets[0].js['user']['name']
    u'Ellen DeGeneres'
    """
    for js in jsons:
        yield Tweet(json.loads(js))


class User(object):
    def __init__(self, jsons):
        self.tweets = [Tweet(t) for t in jsons]


test_tweet = r'''{"favorited": false, "contributors": null, "truncated": false, "text": "If only Bradley's arm was longer. Best photo ever. #oscars http://t.co/C9U5NOtGap", "possibly_sensitive": false, "in_reply_to_status_id": null, "user": {"follow_request_sent": false, "profile_use_background_image": true, "default_profile_image": false, "id": 15846407, "profile_background_image_url_https": "https://pbs.twimg.com/profile_background_images/446411201413541888/uY6CCUHT.png", "verified": true, "profile_text_color": "000000", "profile_image_url_https": "https://pbs.twimg.com/profile_images/421402074153091072/fL_dFiNk_normal.jpeg", "profile_sidebar_fill_color": "C4C6CC", "entities": {"url": {"urls": [{"url": "http://t.co/rJNCGU9GhH", "indices": [0, 22], "expanded_url": "http://www.ellentv.com", "display_url": "ellentv.com"}]}, "description": {"urls": []}}, "followers_count": 28192772, "profile_sidebar_border_color": "FFFFFF", "id_str": "15846407", "profile_background_color": "3ABEE3", "listed_count": 99361, "is_translation_enabled": true, "utc_offset": -25200, "statuses_count": 8635, "description": "Comedian, talk show host and ice road trucker. My tweets are real, and they\\u2019re spectacular.", "friends_count": 46462, "location": "California", "profile_link_color": "3ABEE3", "profile_image_url": "http://pbs.twimg.com/profile_images/421402074153091072/fL_dFiNk_normal.jpeg", "following": false, "geo_enabled": false, "profile_banner_url": "https://pbs.twimg.com/profile_banners/15846407/1395265968", "profile_background_image_url": "http://pbs.twimg.com/profile_background_images/446411201413541888/uY6CCUHT.png", "screen_name": "TheEllenShow", "lang": "en", "profile_background_tile": false, "favourites_count": 160, "name": "Ellen DeGeneres", "notifications": false, "url": "http://t.co/rJNCGU9GhH", "created_at": "Thu Aug 14 03:50:42 +0000 2008", "contributors_enabled": false, "time_zone": "Pacific Time (US & Canada)", "protected": false, "default_profile": false, "is_translator": false}, "geo": null, "id": 440322224407314432, "favorite_count": 1998872, "lang": "en", "entities": {"symbols": [], "user_mentions": [], "hashtags": [{"indices": [51, 58], "text": "oscars"}], "urls": [], "media": [{"expanded_url": "http://twitter.com/TheEllenShow/status/440322224407314432/photo/1", "display_url": "pic.twitter.com/C9U5NOtGap", "url": "http://t.co/C9U5NOtGap", "media_url_https": "https://pbs.twimg.com/media/BhxWutnCEAAtEQ6.jpg", "id_str": "440322224092745728", "sizes": {"small": {"h": 191, "resize": "fit", "w": 340}, "large": {"h": 576, "resize": "fit", "w": 1024}, "medium": {"h": 338, "resize": "fit", "w": 600}, "thumb": {"h": 150, "resize": "crop", "w": 150}}, "indices": [59, 81], "type": "photo", "id": 440322224092745728, "media_url": "http://pbs.twimg.com/media/BhxWutnCEAAtEQ6.jpg"}]}, "created_at": "Mon Mar 03 03:06:13 +0000 2014", "retweeted": false, "coordinates": null, "in_reply_to_user_id_str": null, "in_reply_to_status_id_str": null, "in_reply_to_screen_name": null, "id_str": "440322224407314432", "place": null, "retweet_count": 3422680, "in_reply_to_user_id": null}'''

















twutil/preprocess.py
from collections import defaultdict
import json
import re
import string


punc_re = '[' + re.escape(string.punctuation) + ']'


class Tokenizer(object):
    """
    >>> t = Tokenizer()
    >>> t.tokenize(Tweet(json.loads(data.test_tweet)))
    {'text': [u'if', u'only', u"bradley's", u'arm', u'was', u'longer', u'.', u'best', u'photo', u'ever', u'.', u'HASHTAG_oscars', u'THIS_IS_A_URL'], 'user.name': [u'ellen', u'degeneres']}
    """
    def __init__(self, fields=['text', 'user.name'], ngrams=1, lc=True, collapse_hashtags=False,
                 collapse_mentions=True, collapse_urls=True, limit_repeats=True, retain_punc_toks=True,
                 collapse_digits=False, rt_prefix=False):
        self.ngrams = ngrams
        if ngrams > 1:
            raise NotImplementedError('ngrams > 1 not yet implemented')
        self.fields = fields
        self.lc = lc
        self.collapse_hashtags = collapse_hashtags
        self.collapse_mentions = collapse_mentions
        self.collapse_urls = collapse_urls
        self.limit_repeats = True
        self.retain_punc_toks = retain_punc_toks
        self.collapse_digits = collapse_digits
        self.rt_prefix = rt_prefix

    def do_tokenize(self, text):
        """
        >>> tk = Tokenizer(ngrams=1, lc=True, collapse_urls=True, collapse_mentions=True, collapse_hashtags=False, retain_punc_toks=True)
        >>> tk.do_tokenize('http://www.foo.com fast-forward hi there :) how?? U.S.A. @you whaaaaaaat? #awesome.')
        ['THIS_IS_A_URL', 'fast-forward', 'hi', 'there', ':)', 'how', '??', 'u.s.a', '.', 'THIS_IS_A_MENTION', 'what', '?', 'HASHTAG_awesome', '.']
        >>> tk = Tokenizer(rt_prefix=True)
        >>> tk.do_tokenize('RT hi there')
        ['RT_hi', 'RT_there']
        >>> tk.do_tokenize('hi there')
        ['hi', 'there']
        """
        text = text.lower() if self.lc else text
        if self.collapse_hashtags:
            text = re.sub('#\S+', 'THIS_IS_A_HASHTAG', text)
        else:
            text = re.sub('#(\S+)', r'HASHTAG_\1', text)
        if self.collapse_mentions:
            text = re.sub('@\S+', 'THIS_IS_A_MENTION', text)
        if self.collapse_urls:
            text = re.sub('http\S+', 'THIS_IS_A_URL', text)
        if self.limit_repeats:
            text = re.sub(r'(.)\1\1\1+', r'\1', text)
        if self.collapse_digits:
            text = re.sub(r'[0-9]+', '9', text)
        toks = []
        for tok in text.split():
            tok = re.sub(r'^(' + punc_re + '+)', r'\1 ', tok)
            tok = re.sub(r'(' + punc_re + '+)$', r' \1', tok)
            for subtok in tok.split():
                if self.retain_punc_toks or re.search('\w', subtok):
                    toks.append(subtok)
        if self.rt_prefix:
            rt_text = 'rt' if self.lc else 'RT'
            if rt_text in toks:
                toks.remove(rt_text)
                toks = ['RT_' + t for t in toks]
        return toks

    def tokenize(self, tw):
        """ Given a Tweet object, return a dict mapping field name to tokens. """
        toks = defaultdict(lambda: [])
        for field in self.fields:
            if '.' in field:
                parts = field.split('.')
                value = tw.js
                for p in parts:
                    value = value[p]
            else:
                value = tw.js[field]
            toks[field].extend(self.do_tokenize(value))
        return dict([(f, v) for f, v in toks.iteritems()])


twutil/collect.py
import json
import os
import sys
from threading import Thread
import time
import traceback
from TwitterAPI import TwitterAPI

try:
    from Queue import Queue
except:
    from queue import Queue

twapi = None


def reinit():
    global twapi
    twapi = TwitterAPI(os.environ.get('TW_CONSUMER_KEY'),
                       os.environ.get('TW_CONSUMER_SECRET'),
                       os.environ.get('TW_ACCESS_TOKEN'),
                       os.environ.get('TW_ACCESS_TOKEN_SECRET'))

reinit()


def list_members(slug, owner_screen_name, limit=1e10):
    """See https://dev.twitter.com/rest/reference/get/lists/members"""
    cursor = -1
    members = []
    try:
        response = twapi.request('lists/members', {'slug': slug, 'owner_screen_name': owner_screen_name, 'cursor': cursor, 'count': 5000})
        if response.status_code in [88, 130, 420, 429]:  # rate limit
            sys.stderr.write('Error for %s/%s: %s\nSleeping for 5 minutes...\n' % (owner_screen_name, slug, response.text))
            time.sleep(301)
        elif response.status_code != 200:
            sys.stderr.write('Skipping bad query: %s\n' % response.text)
            return members
        else:
            items = [r['screen_name'] for r in response if 'screen_name' in r]
            return items
    except Exception as e:
        sys.stderr.write('Error: %s\nskipping...\n' % e)
        return members
    return members


def lookup_handles(ids):
    """ Fetch the twitter screen_names of each id. """
    names = set()
    for id_list in [ids[100 * i:100 * i + 100] for i in range(len(ids))]:
        if len(id_list) > 0:
            while True:
                r = twapi.request('users/lookup', {'user_id': ','.join([str(i) for i in id_list])})
                if r.status_code in [88, 130, 420, 429]:  # rate limit
                    sys.stderr.write('Sleeping off rate limit for %s: %s\n' % (str(id_list), r.text))
                    time.sleep(301)
                elif r.status_code == 200:
                    for item in r.get_iterator():
                        names.add((item['screen_name'], item['id_str']))
                    break
                else:
                    sys.stderr.write('Error: %s\nSkipping %s...\n' % (str(id_list), r.text))
                    break

    return names


def lookup_ids(handles):
    """ Fetch the twitter ids of each screen_name. """
    ids = set()
    for handle_list in [handles[100 * i:100 * i + 100] for i in range(len(handles))]:
        if len(handle_list) > 0:
            while True:
                r = twapi.request('users/lookup', {'screen_name': ','.join(handle_list)})
                if r.status_code in [88, 130, 420, 429]:  # rate limit
                    sys.stderr.write('Sleeping off rate limit for %s: %s\n' % (str(handle_list), r.text))
                    time.sleep(300)
                elif r.status_code == 200:
                    for item in r.get_iterator():
                        ids.add(item['id_str'])
                    break
                else:
                    sys.stderr.write('Error: %s\nSkipping %s...\n' % (str(handle_list), r.text))
                    break

    return ids


def tweets_for_id(user_id, limit=1e10):
    """ Collect the most recent 3200 tweets for this user_id, sleeping to deal with rate limits."""
    # Map id to screen_name
    r = twapi.request('users/lookup', {'user_id': user_id})
    if r.status_code == 200:
        sname = [t for t in r][0]['screen_name']
        return tweets_for_user(sname, limit)
    else:
        sys.stderr.write('error:' % r.text)


def tweets_for_user(screen_name, limit=1e10):
    """ Collect the most recent 3200 tweets for this user, sleeping to deal with rate limits."""
    qu = Queue()
    p = Thread(target=_tweets_for_user, args=(qu, screen_name, limit))
    p.start()
    p.join(910)
    if p.is_alive():
        sys.stderr.write('no results after 15 minutes for %s. Aborting.' % screen_name)
        return []
    else:
        return qu.get()


def _tweets_for_user(qu, screen_name, limit=1e10):
    max_id = None
    tweets = []
    while True:
        # TODO: Have to kill threads after some time; otherwise, we make too many...
        try:
            if max_id:
                response = twapi.request('statuses/user_timeline', {'screen_name': screen_name, 'count': 200, 'max_id': max_id})
            else:
                response = twapi.request('statuses/user_timeline', {'screen_name': screen_name, 'count': 200})
            if response.status_code == 34 or response.status_code == 404 or response.status_code == 401:
                sys.stderr.write('Skipping bad user: %s\n' % response.text)
                qu.put(tweets)
                return
            elif response.status_code != 200:  # something went wrong
                sys.stderr.write('Error: %s\nSleeping for 5 minutes...\n' % response.text)
                time.sleep(300)
            else:
                items = [t for t in response]
                if len(items) == 0:
                    qu.put(tweets)
                    return
                else:
                    sys.stderr.write('fetched %d more tweets for %s\n' % (len(items), screen_name))
                    tweets.extend(items)
                    if len(tweets) >= limit:
                        qu.put(tweets[:limit])
                        return
                max_id = min(t['id'] for t in response) - 1
        except Exception as e:
            sys.stderr.write('Error: %s Skipping...\n' % e)


def track_user_ids(ids):
    """ Return an iterator tweets from users in this id list. """
    results = twapi.request('statuses/filter', {'follow': ','.join(ids)})
    return results.get_iterator()


def friends_for_id(id_, limit=1e10):
    # FIXME: DRY from _tweets_for_user
    cursor = -1
    friends = []
    while len(friends) < limit:
        try:
            response = twapi.request('friends/ids', {'user_id': id_, 'count': 5000,
                                                     'cursor': cursor, 'stringify_ids': True})
            if response.status_code in [88, 130, 420, 429]:  # rate limit
                sys.stderr.write('Error for %s: %s\nSleeping for 5 minutes...\n' % (id_, response.text))
                time.sleep(300)
            elif response.status_code != 200:
                sys.stderr.write('Skipping bad user: %s\n' % response.text)
                return friends
            else:
                response = json.loads(response.text)
                items = [r for r in response['ids']]
                if len(items) == 0:
                    return friends
                else:
                    sys.stderr.write('fetched %d more friends for %s\n' % (len(items), id_))
                    time.sleep(1)
                    friends.extend(items)
                    if len(friends) >= limit:
                        return friends[:limit]
                cursor = response['next_cursor']
        except Exception as e:
            sys.stderr.write('Error: %s\nskipping...\n' % e)
            sys.stderr.write(traceback.format_exc())
            return friends
    return friends


def friends_for_user(screen_name, limit=5000):
    id_ = [i for i in lookup_ids([screen_name])]
    if len(id_) == 1:
        return friends_for_id(id_, limit)
    else:
        sys.stderr.write('cannot find id for user %s' % screen_name)
        return []


def get_friends(id_=None, screen_name=None, limit=1e10):
    """ Either id_ or screen_name must not be None. """
    # FIXME: DRY from _tweets_for_user
    if not (id_ or screen_name):
        raise Exception("either id_ or screen_name must not be None")
    if id_:
        key = 'user_id'
        val = id_
    else:
        key = 'screen_name'
        val = screen_name

    cursor = -1
    friends = []
    while len(friends) < limit:
        try:
            response = twapi.request('friends/ids', {key: val, 'count': 5000,
                                                     'cursor': cursor, 'stringify_ids': True})
            if response.status_code in [88, 130, 420, 429]:  # rate limit
                sys.stderr.write('Error for %s: %s\nSleeping for 5 minutes...\n' % (val, response.text))
                time.sleep(300)
            elif response.status_code != 200:
                sys.stderr.write('Skipping bad user: %s\n' % response.text)
                return friends
            else:
                response = json.loads(response.text)
                items = [r for r in response['ids']]
                if len(items) == 0:
                    return friends
                else:
                    sys.stderr.write('fetched %d more friends for %s\n' % (len(items), val))
                    time.sleep(1)
                    friends.extend(items)
                    if len(friends) >= limit:
                        return friends[:limit]
                cursor = response['next_cursor']
        except Exception as e:
            sys.stderr.write('Error: %s\nskipping...\n' % e)
            sys.stderr.write(traceback.format_exc())
            return friends
    return friends


def followers_for_user(screen_name, limit=1e10):
    id_ = [i for i in lookup_ids([screen_name])]
    if len(id_) == 1:
        return followers_for_id(id_, limit)
    else:
        sys.stderr.write('cannot find id for user %s' % screen_name)
        return []


def get_followers(id_=None, screen_name=None, limit=1e10):
    """ Either id_ or screen_name must not be None. """
    # FIXME: DRY from _tweets_for_user
    if not (id_ or screen_name):
        raise Exception("either id_ or screen_name must not be None")
    if id_:
        key = 'user_id'
        val = id_
    else:
        key = 'screen_name'
        val = screen_name
    cursor = -1
    followers = []
    while len(followers) < limit:
        try:
            response = twapi.request('followers/ids', {key: val, 'count': 5000,
                                                       'cursor': cursor, 'stringify_ids': True})
            if response.status_code in [88, 130, 420, 429]:  # rate limit
                sys.stderr.write('Error for %s: %s\nSleeping for 5 minutes...\n' % (val, response.text))
                time.sleep(300)
            elif response.status_code != 200:
                sys.stderr.write('Skipping bad user: %s\n' % response.text)
                return followers
            else:
                result = json.loads(response.text)
                items = [r for r in response]
                if len(items) == 0:
                    return followers
                else:
                    sys.stderr.write('fetched %d more followers for %s\n' % (len(items), val))
                    time.sleep(1)
                    followers.extend(items)
                    if len(followers) >= limit:
                        return followers[:limit]
                cursor = result['next_cursor']
        except Exception as e:
            sys.stderr.write('Error: %s\nskipping...\n' % e)
            sys.stderr.write(traceback.format_exc())
            return followers
    return followers


def followers_for_id(id_, limit=1e10):
    return get_followers(id_=id_, limit=limit)


def followers_for_screen_name(screen_name, limit=1e10):
    return get_followers(screen_name=screen_name, limit=limit)
